src/moin/storage/middleware/protecting.py

```diff
@@ -340,6 +340,14 @@ def __init__(self, protector, rev, p_item=None):
         self.rev = rev
         self.item = p_item or ProtectedItem(protector, rev.item)
 
+    def __str__(self):
+        """
+        Return revision contents as a string.
+        """
+        self.data.seek(0)
```

Comment:
that looks strange. where does self.data come from?

did you mean `self.rev.data`?

also, this does not really belong into the protecting middleware, so i guess here there should be only a stub that delegates to `rev.__str__` and the implementation should be there.

Reply:
also, as this is just another way to get the data, guess we need the same permission checks as in the `data` property, see below.

Reply:
self is a ProtectedRevision object. I think it is best to revert 4024102 and 708ec5d and reopen #890. 

I do not understand what you meant by best = "text = str(rev): in the #890 issue.

Also, seems there is very little code saved.

Reply:
I meant implementing `__str__`.

Just do not put code that has nothing to do with "protection" (== permissions/acl checking) into the protection middleware.

That functionality is rather fundamental and thus should be implemented in the Revision class.
And of course there also needs to be a __str__ in the ProtectedRevision, but that would just check permisssions and then forward to the unprotected Revision.

Reply:
ProtectedRevision already has:

    @property
    def meta(self):
        self.require(READ, PUBREAD)
        return self.rev.meta

    @property
    def data(self):
        self.require(READ, PUBREAD)
        return self.rev.data

The above seem to get called when .data is referenced by a call to the `__str__`  method, so no need to check again.

Reply:
You would not call these, but `str(self.rev)` - and before doing that, check permissions (only once).